pyccel/parser/syntactic.py

```diff
@@ -823,12 +821,12 @@ def fill_types(ls):
         results = []
         result_counter = 1
         for i in zip(*returns):
-            if not all(i[0]==j for j in i) or not isinstance(i[0], Symbol):
+            if not all(i[0]==j for j in i) or not isinstance(i[0], PyccelSymbol):
                 result_name, result_counter = create_variable(self._used_names,
                                                               prefix = 'Out',
                                                               counter = result_counter)
                 results.append(result_name)
-            elif isinstance(i[0], Symbol) and any(i[0].name==x.name for x in arguments):
+            elif isinstance(i[0], PyccelSymbol) and any(i[0]==x.name for x in arguments):
                 result_name, result_counter = create_variable(self._used_names,
                                                               prefix = 'Out',
                                                               counter = result_counter)
```

Comment:
We should avoid calling `i.__getitem__(0)` multiple times in this `if` block. This can be done by simply storing `i0 = i[0]` at the beginning. Further, there is a lot of duplicated code here, and it can be cleaned up:

```python
for r in zip(*returns):
    r0 = r[0]

    pyccel_symbol  = isinstance(r0, PyccelSymbol)
    same_results   = all(r0 == ri for ri in r)
    name_available = all(r0 != a.name for a in arguments)

    if pyccel_symbol and same_results and name_available:
        result_name = r0
    else:
        result_name, result_counter = create_variable(self._used_names, \
                                prefix = 'Out', counter = result_counter)

    results.append(result_name)
```

Reply:
049c142